README.md
PyNamecoin
===========
A Python wrapper for the Namecoin (NMC) API, using the standard Namecoin-client.

Installation
-----------
There are two ways to install this package, pip or manually. Preferably use the simple pip-installation, incredibly easy. 
* pip:
```
pip install pynamecoin
```
* custom:
	1. Download one of the recent releases on the Releases-page.
	2. Move this folder to your Python installation, under /plugins/
	3. Done!

Usage
-----------
1. Run a server-instance of namecoind.exe. 
The first instance without any args will automatically be run as a server. 
A client is provided under <installation-folder>/pynamecoin/bin/runserver.py.
2. Wait a couple of minutes for the server to synchronize with the network. Note: No feedback is given in the commandline by the namecoind.exe.
3. Execute your script. For example: <installation-folder>/pynamecoin/bin/simple_example.py

Documentation
-----------
<TODO>

License
-----------
This software is free to use, any where, any time and in any way.
However, I would like to know in what projects this software ends up; please send me a mail. 

Any coffee-compensation can be send to:
* BTC: 1E8Ao51Y9UNKvRus17yKEqXPj6fnAeMQYb
* NMC: NFsoyZ8uxj44M3ANgqVsPepvUhHBKRiiyG

CHANGES.md
v1.0.0, 12/3/2014 -- Initial release.

pynamecoin/test/test_info.py
import unittest
from .. import info;

# namecoind.exe server needs to running.
class TestCore(unittest.TestCase):
	
	def test_validate_true(self):
		val = info.validate('MyJaz1tvfjHby4ga7DkR6WbquDVy1aV5Xc');
		self.assertTrue(val)
	
	def test_validate_false(self):
		val = info.validate('42');
		self.assertFalse(val);
		
	def test_validate_none(self):
		val = info.validate(None);
		self.assertFalse(val);
		
	def test_getinfo(self):
		val = info.getinfo()
		self.assertEqual(type(val), dict);
		self.assertEqual(len(val), 16);
	
	def test_getblockcount(self):
		control = info.getinfo()
		val = info.getblockcount();
		self.assertEqual(type(val), int);
		self.assertEqual(control['blocks'], val);
	
	def test_getdifficulty(self):
		control = info.getinfo()
		val = info.getdifficulty();
		self.assertEqual(type(val), float);
		self.assertEqual(control['difficulty'], val);
	
	def test_gethashespersec(self):
		control = info.getinfo()
		val = info.gethashespersec();
		self.assertEqual(type(val), int);
		self.assertEqual(control['hashespersec'], val);
	
	def test_getconnectioncount(self):
		control = info.getinfo()
		val = info.getconnectioncount();
		self.assertEqual(type(val), int);
		self.assertEqual(control['connections'], val);
		
	
	def test_name_show_exists(self):
		name = 'd/domain'
		val = info.name_show(name)
		self.assertEqual(type(val), dict);
		self.assertEqual(val['name'], name);
		
	def test_name_show_not_existing(self):
		name = 'dadajjunonexisting'
		val = info.name_show(name)
		self.assertEquals(val, None);
		
	def test_name_show_None(self):
		name = None
		val = info.name_show(name)
		self.assertEquals(val, None); # raise exception?
	
	def test_name_history_exists(self):
		name = 'd/domain';
		val = info.name_history(name);
		self.assertEqual(type(val), list);
		self.assertEqual(type(val[0]), dict);
		
	def test_name_history_not_existing(self):
		name = 'dadajjunonexisting'
		val = info.name_show(name)
		self.assertEquals(val, None);
		
	def test_name_history_none(self):
		name = None;
		val = info.name_history(name);
		self.assertEquals(val, None);
	
	def test_name_scan_exists(self):
		start_name = 'd/domain';
		count = 10;
		val = info.name_scan(start_name, count);
		self.assertEqual(type(val), list);
		self.assertEqual(len(val), count);
		self.assertEqual(type(val[0]), dict);
	
	def test_name_scan_not_existing(self):
		# Should this return results?
		start_name = 'saksjfsduiaehuh';
		count = 10;
		val = info.name_scan(start_name, count);
		self.assertEqual(type(val), list);
		self.assertEqual(len(val), count);
		self.assertEqual(type(val[0]), dict);
	
	def test_name_scan_none(self):
		with self.assertRaises(Exception):
			start_name = None;
			count = 10;
			val = info.name_scan(start_name, count);
		
	def test_name_scan_invaid_count(self):
		with self.assertRaises(Exception):
			start_name = 'd/domain'
			count = 0 
			val = info.name_scan(start_name, count);
	
	def test_name_filter(self):
		regex = 'test'
		val = info.name_filter(regex, count=10);
		self.assertEquals(type(val), list);
		self.assertEqual(type(val[0]), dict);
		
	def test_name_filter_invalid_regex(self):
		with self.assertRaises(Exception):
			regex = '\p**\\\\#\@'
			val = info.name_filter(regex, count=10);
	
	def test_name_filter_invalid_count(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = info.name_filter(regex, count=-1);
		
	def test_name_filter_invalid_maxage(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = info.name_filter(regex, maxage=-1, minage=-10);
	
	def test_name_filter_invalid_minage(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = info.name_filter(regex, minage=-1);
	
	def test_name_filter_conflicting_ages(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = info.name_filter(regex, maxage=100, minage=500);
	
	def test_name_filter_same_age(self):
		regex = 'test'
		val = info.name_filter(regex, maxage=10, minage=10);
		self.assertEquals(val, []);
	
	'''
	def test_gettransaction(self):
		val = info.gettransaction('6cce69f22841439b40bd576d15c46ec6d60ad44e1d604b51049ecebd011b53d4');
		self.assertEquals(val, None); 
	'''
	
	def test_gettransaction_nonexisting(self):
		with self.assertRaises(Exception):
			txid = '42';
			val = info.gettransaction(txid);
			self.assertEquals(val, None); 
	
	def test_gettransaction_none(self):
		with self.assertRaises(Exception):
			txid = None;
			val = info.gettransaction(txid);
			
if __name__ == '__main__':
    unittest.main()

core.py
import subprocess, json, logging
'''
Settings
'''
CLIENT  = 'namecoin/namecoind.exe'
LOG 	= 'pynamecoin.log'
SAFE_MODE = True

'''
Logger
'''
FORMAT = '%(asctime)-15s %(levelname)-8s: %(message)s'
logging.basicConfig(format=FORMAT, level=logging.INFO, filename=LOG)
LOGGER = logging.getLogger('pynmc')


'''
ServerException is issued when server returns errors
'''
class ServerException(Exception):
	def __init__(self, value, code=0):
		LOGGER.error(value);
		self.value = value;
		self.code = code
	
	def __str__(self):
		return repr(self.value) + "(" + str(self.code) + ')';

'''
Pair dataclass
'''
class Pair:
	def __init__(self, name, value, owner_address, txid, expires_in):
		self.name = name;
		self.value = value;
		self.owner_address = owner_address;
		self.txid = txid;
		self.expires_in = expires_in;
		
	def expired():
		return expires_in < 0;
	
	def __str__():
		return str(self.name) + ": " + repr(self.value);

'''
Retrieve the transaction with id txid.
'''
def gettransaction(txid):
	value = nmc_call('gettransaction', [str(txid)]);
	return value

'''
Return general information about the state of the network. Contains:
{version, balance, blocks, timeoffset, connections, proxy, 
	generate, genproclimit, difficulty, hashespersec, testnet,keypoololdest,
	keypoolsize,paytxfee,mininput,errors}
'''
def getinfo():
	return nmc_call('getinfo');


'''
Get the current block number
'''
def getblockcount():
	return nmc_call('getblockcount');

'''
Get the current mining difficulty
'''
def getdifficulty():
	return nmc_call('getdifficulty');

'''
Get the current hashrate of the network
'''	
def gethashespersec():
	return nmc_call('gethashespersec');

'''
Get connection count of the node
'''
def getconnectioncount():
	return nmc_call('getconnectioncount');
	
'''
Validate NMC address
'''
def validate(address):
	if type(address) is str:
		value = nmc_call('validateaddress', [address]);
		return value['isvalid'] == True;
	else:
		return False;
'''
Scans n number of names alphabetically, starting at start_name
'''
def name_scan(start_name, count):
	assert count > 0
	return nmc_call('name_scan', [(start_name), str(count)]);

'''
Retrieve names matching the regex (and additional parameters
'''
def name_filter(regex, maxage=36000, minage=0, count=0, stat = False):
	assert minage <= maxage;
	assert minage >= 0;
	assert maxage >= 0;
	assert count >= 0;
	stat = 'stat' if stat else None;
	value = nmc_call('name_filter', [(regex), str(maxage), str(minage), str(count)]);
	return value
	
'''
Return the associated value of the <name>, otherwise return None.
'''
def name_show(name):
	try:
		value = nmc_call('name_show',[(name)]);
	except:
		value = None;
	return value;

'''
Return the history (value updates) of the <name>.
'''
def name_history(name):
	try:
		value = nmc_call('name_history',[(name)]);
	except:
		value = None;
	return value;

'''
Calls the client with the method and additional args.
'''
def nmc_call(method, args = []):
	rargs = [CLIENT,method] + args;
	p = subprocess.Popen(rargs,stdout=subprocess.PIPE,stderr=subprocess.PIPE);
	out, err = p.communicate();
	LOGGER.info('input: ' + ' '.join(rargs));
	if err:
		raise ServerException(str(err['message']), str(err['code']));
	try:
		ret =  json.loads(out.replace('\r','').replace('\n',''));
	except Exception as e:
		raise ServerException(str(e) + ": " + repr(out));	
	return ret

pynamecoin/info.py
import subprocess, json, logging
'''
Settings
'''
CLIENT  = './bin/namecoin/namecoind.exe'

'''
Logger
'''
FORMAT 	= '%(asctime)-15s %(levelname)-8s: %(message)s'
LOG 	= 'pynamecoin.log'
logging.basicConfig(format=FORMAT, level=logging.INFO, filename=LOG)
LOGGER 	= logging.getLogger('pynmc')


'''
ServerException is issued when server returns errors
'''
class ServerException(Exception):
	def __init__(self, value, code=0):
		LOGGER.error(value);
		self.value = value;
		self.code = code
	
	def __str__(self):
		return repr(self.value) + "(" + str(self.code) + ')';

'''
Pair dataclass
'''
class Pair:
	def __init__(self, **fields):
		self.__dict__.update(fields);
		
	def expired():
		return expires_in < 0;
	
	def __str__():
		return str(self.name) + ": " + repr(self.value);

'''
Retrieve the transaction with id txid.
'''
def gettransaction(txid):
	value = nmc_call('gettransaction', [str(txid)]);
	return value

'''
Return general information about the state of the network. Contains:
{version, balance, blocks, timeoffset, connections, proxy, 
	generate, genproclimit, difficulty, hashespersec, testnet,keypoololdest,
	keypoolsize,paytxfee,mininput,errors}
'''
def getinfo():
	return nmc_call('getinfo');


'''
Get the current block number
'''
def getblockcount():
	return nmc_call('getblockcount');

'''
Get the current mining difficulty
'''
def getdifficulty():
	return nmc_call('getdifficulty');

'''
Get the current hashrate of the network
'''	
def gethashespersec():
	return nmc_call('gethashespersec');

'''
Get connection count of the node
'''
def getconnectioncount():
	return nmc_call('getconnectioncount');
	
'''
Validate NMC address
'''
def validate(address):
	if type(address) is str:
		value = nmc_call('validateaddress', [address]);
		return value['isvalid'] == True;
	else:
		return False;
'''
Scans n number of names alphabetically, starting at start_name
'''
def name_scan(start_name, count):
	assert count > 0
	return nmc_call('name_scan', [(start_name), str(count)]);

'''
Retrieve names matching the regex (and additional parameters
'''
def name_filter(regex, maxage=36000, minage=0, count=0, stat = False):
	assert minage <= maxage;
	assert minage >= 0;
	assert maxage >= 0;
	assert count >= 0;
	stat = 'stat' if stat else None;
	value = nmc_call('name_filter', [(regex), str(maxage), str(minage), str(count)]);
	return value
	
'''
Return the associated value of the <name>, otherwise return None.
'''
def name_show(name):
	try:
		value = nmc_call('name_show',[(name)]);
	except:
		value = None;
	return Pair(**value);

'''
Return the history (value updates) of the <name>.
'''
def name_history(name):
	try:
		value = nmc_call('name_history',[(name)]);
	except:
		value = None;
	return value;

'''
Calls the client with the method and additional args.
'''
def nmc_call(method, args = []):
	rargs = [CLIENT,method] + args;
	p = subprocess.Popen(rargs,stdout=subprocess.PIPE,stderr=subprocess.PIPE);
	out, err = p.communicate();
	LOGGER.info('input: ' + ' '.join(rargs));
	if err:
		err = err[7:];
		jerr =  json.loads(err.replace('\r','').replace('\n',''));
		raise ServerException(str(jerr['message']), str(jerr['code']));
	try:
		ret =  json.loads(out.replace('\r','').replace('\n',''));
	except Exception as e:
		raise ServerException(str(e) + ": " + repr(out));	
	return ret

testcore.py
import unittest
import core

class TestCore(unittest.TestCase):
	
	def test_validate_true(self):
		val = core.validate('MyJaz1tvfjHby4ga7DkR6WbquDVy1aV5Xc');
		self.assertTrue(val)
	
	def test_validate_false(self):
		val = core.validate('42');
		self.assertFalse(val);
		
	def test_validate_none(self):
		val = core.validate(None);
		self.assertFalse(val);
		
	def test_getinfo(self):
		val = core.getinfo()
		self.assertEqual(type(val), dict);
		self.assertEqual(len(val), 16);
	
	def test_getblockcount(self):
		control = core.getinfo()
		val = core.getblockcount();
		self.assertEqual(type(val), int);
		self.assertEqual(control['blocks'], val);
	
	def test_getdifficulty(self):
		control = core.getinfo()
		val = core.getdifficulty();
		self.assertEqual(type(val), float);
		self.assertEqual(control['difficulty'], val);
	
	def test_gethashespersec(self):
		control = core.getinfo()
		val = core.gethashespersec();
		self.assertEqual(type(val), int);
		self.assertEqual(control['hashespersec'], val);
	
	def test_getconnectioncount(self):
		control = core.getinfo()
		val = core.getconnectioncount();
		self.assertEqual(type(val), int);
		self.assertEqual(control['connections'], val);
		
	
	def test_name_show_exists(self):
		name = 'd/domain'
		val = core.name_show(name)
		self.assertEqual(type(val), dict);
		self.assertEqual(val['name'], name);
		
	def test_name_show_not_existing(self):
		name = 'dadajjunonexisting'
		val = core.name_show(name)
		self.assertEquals(val, None);
		
	def test_name_show_None(self):
		name = None
		val = core.name_show(name)
		self.assertEquals(val, None); # raise exception?
	
	def test_name_history_exists(self):
		name = 'd/domain';
		val = core.name_history(name);
		self.assertEqual(type(val), list);
		self.assertEqual(type(val[0]), dict);
		
	def test_name_history_not_existing(self):
		name = 'dadajjunonexisting'
		val = core.name_show(name)
		self.assertEquals(val, None);
		
	def test_name_history_none(self):
		name = None;
		val = core.name_history(name);
		self.assertEquals(val, None);
	
	def test_name_scan_exists(self):
		start_name = 'd/domain';
		count = 10;
		val = core.name_scan(start_name, count);
		self.assertEqual(type(val), list);
		self.assertEqual(len(val), count);
		self.assertEqual(type(val[0]), dict);
	
	def test_name_scan_not_existing(self):
		# Should this return results?
		start_name = 'saksjfsduiaehuh';
		count = 10;
		val = core.name_scan(start_name, count);
		self.assertEqual(type(val), list);
		self.assertEqual(len(val), count);
		self.assertEqual(type(val[0]), dict);
	
	def test_name_scan_none(self):
		with self.assertRaises(Exception):
			start_name = None;
			count = 10;
			val = core.name_scan(start_name, count);
		
	def test_name_scan_invaid_count(self):
		with self.assertRaises(Exception):
			start_name = 'd/domain'
			count = 0 
			val = core.name_scan(start_name, count);
	
	def test_name_filter(self):
		regex = 'test'
		val = core.name_filter(regex, count=10);
		self.assertEquals(type(val), list);
		self.assertEqual(type(val[0]), dict);
		
	def test_name_filter_invalid_regex(self):
		with self.assertRaises(Exception):
			regex = '\p**\\\\#\@'
			val = core.name_filter(regex, count=10);
	
	def test_name_filter_invalid_count(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = core.name_filter(regex, count=-1);
		
	def test_name_filter_invalid_maxage(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = core.name_filter(regex, maxage=-1, minage=-10);
	
	def test_name_filter_invalid_minage(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = core.name_filter(regex, minage=-1);
	
	def test_name_filter_conflicting_ages(self):
		with self.assertRaises(Exception):
			regex = 'test'
			val = core.name_filter(regex, maxage=100, minage=500);
	
	def test_name_filter_same_age(self):
		regex = 'test'
		val = core.name_filter(regex, maxage=10, minage=10);
		self.assertEquals(val, []);
	
	'''
	def test_gettransaction(self):
		val = core.gettransaction('6cce69f22841439b40bd576d15c46ec6d60ad44e1d604b51049ecebd011b53d4');
		self.assertEquals(val, None); 
	'''
	
	def test_gettransaction_nonexisting(self):
		with self.assertRaises(Exception):
			txid = '42';
			val = core.gettransaction(txid);
			self.assertEquals(val, None); 
	
	def test_gettransaction_none(self):
		with self.assertRaises(Exception):
			txid = None;
			val = core.gettransaction(txid);
			
if __name__ == '__main__':
    unittest.main()